### ISSUE TITLE
ndiffports is not working. Second sub flow is not created.

### ISSUE THREAD

**Issue by Scorpio1987** [multipath-tcp/mptcp#159]
I am trying to test mptcp in my machine. I have configured path manager as "ndiffports" but i don't see the second sub flow is created. After enabling the mptcp debugs i see that the system is trying to create second subflow but it is failing. Here is the logs.

[22171.387670] net/mptcp/mptcp_ctrl.c: mptcp_alloc_mpcb: created mpcb with token 0x507e83fb
[22171.387724] net/mptcp/mptcp_ctrl.c: mptcp_add_sock: token 0x507e83fb pi 1, src_addr:16.0.0.1:37988 dst_addr:15.0.0.1:22, cnt_subflows now 1
[22171.389248] net/mptcp/mptcp_ctrl.c: mptcp_add_sock: token 0x507e83fb pi 2, src_addr:0.0.0.0:0 dst_addr:0.0.0.0:0, cnt_subflows now 2
[22171.389276] net/mptcp/mptcp_ipv4.c: mptcp_init4_subsockets: token 0x507e83fb pi 2 src_addr:16.0.0.1:0 dst_addr:15.0.0.1:22 ifidx: -30720
[22171.389299] net/mptcp/mptcp_ipv4.c: mptcp_init4_subsockets: MPTCP subsocket connect() failed, error -19
[22171.389326] net/mptcp/mptcp_ctrl.c: mptcp_del_sock: Removing subsock tok 0x507e83fb pi:2 state 7 is_meta? 0
[22177.497744] net/mptcp/mptcp_ctrl.c: mptcp_close: Close of meta_sk with tok 0x507e83fb
[22177.502169] net/mptcp/mptcp_ctrl.c: mptcp_del_sock: Removing subsock tok 0x507e83fb pi:1 state 7 is_meta? 0
[22177.502235] net/mptcp/mptcp_ctrl.c: mptcp_sock_destruct destroying meta-sk

here the error returned is 19 which is ENODEV. This may be because the ifidx shown in the log is invalid "ifidx: -30720".
From the code browsing i found out that create_subflow_worker calls mptcp_init4_subsockets and passes "loc" which is a local variable which is not initialised to 0's. Inside mptcp_init4_subsockets, the code is accessing "loc->if_idx" and setting it to "sk->sk_bound_dev_if". Since "loc" is not initialised this will contain junk value and this may be causing the failure.

**Comment by cpaasch** [comment]
Thanks for reporting this. There is a patch for this, posted by Benjamin Hesmans at https://sympa-2.sipr.ucl.ac.be/sympa/arc/mptcp-dev/2016-12/msg00005.html .

We will apply this fix as soon as mptcp_trunk has moved to v4.4.

**Comment by cpaasch** [comment]
This has now been fixes:

```
commit a2ba2dac2d419c274d86f1221a49f99f004262a6
Author: Benjamin Hesmans <benjamin.hesmans@uclouvain.be>
Date:   Fri Dec 2 08:26:43 2016 +0100

    MPTCP : fix if_idx in ndiffports

    if_idx was not initialized

    Fixes: 9e763119cccc6 (mptcp: Check interface index in fullmesh pm)
    Signed-off-by: Benjamin Hesmans <benjamin.hesmans@uclouvain.be>
    Signed-off-by: Christoph Paasch <cpaasch@apple.com>
```
